state/protocol/kvstore.go

```diff
@@ -0,0 +1,71 @@
+package protocol
+
+import "github.com/onflow/flow-go/model/flow"
+
+// VersionedEncodable defines the interface for a versioned key-value store independent
+// of the set of keys which are supported. This allows the storage layer to support
+// storing different key-value model versions within the same software version.
+type VersionedEncodable interface {
+	// VersionedEncode encodes the key-value store, returning the version separately
+	// from the encoded bytes.
+	// No errors are expected during normal operation.
+	VersionedEncode() (uint64, []byte, error)
+}
+
+// ViewBasedActivator allows setting value that will be active from specific view.
+type ViewBasedActivator[T any] struct {
+	Data           T
+	ActivationView uint64 // first view at which Data will take effect
+}
```

Comment:
I would suggest to move these definitions to the bottom of the file. Or we could move [lines 5- 19](https://github.com/onflow/flow-go/blob/43c3ceb44c015c8c652ebe5ed43b7d3493971801/state/protocol/kvstore.go#L5-L19) below [lines 21- 26](https://github.com/onflow/flow-go/blob/43c3ceb44c015c8c652ebe5ed43b7d3493971801/state/protocol/kvstore.go#L21-L26). 
Ether works, as long as [lines 21- 26](https://github.com/onflow/flow-go/blob/43c3ceb44c015c8c652ebe5ed43b7d3493971801/state/protocol/kvstore.go#L21-L26) are at the top of the file, because these lines are an introductory comment for the entire file.

Reply:
[`3ed352f` (#5650)](https://github.com/onflow/flow-go/pull/5650/commits/3ed352f9d8e4a88fa848b87e902710d832501784)